File: 1_algorithmic_toolbox/week2/python/4_last_digit_sum_fibonacci.py
def last_digit_sum_fibonacci(n):
    MODULO = 10

    cumulated_pisano = [0]
    a, b, sum = 1, 1, 1
    while not (a == 0 and b == 1):
        cumulated_pisano.append(sum)
        a, b = b, (a + b) % MODULO
        sum = (sum + a) % MODULO

    p_sz = len(cumulated_pisano)
    q, r = n // p_sz, n % p_sz
    return (q * cumulated_pisano[p_sz - 1] + cumulated_pisano[r]) % MODULO


if __name__ == "__main__":
    print(last_digit_sum_fibonacci(int(input())))

File: 1_algorithmic_toolbox/week2/rust/4_last_digit_sum_fibonacci.rs
const MODULO: i64 = 10;

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let n = input.trim().parse::<i64>().unwrap();
    println!("{}", last_digit_sum_fibonacci(n));
}

fn last_digit_sum_fibonacci(n: i64) -> i64 {
    let mut cumulated_pisano = vec![0];
    let (mut a, mut b, mut sum) = (1, 1, 1);
    while !(a == 0 && b == 1) {
        cumulated_pisano.push(sum);
        (a, b) = (b, (a + b) % MODULO);
        sum = (sum + a) % MODULO;
    }

    let p_sz = cumulated_pisano.len() as i64;
    let (q, r) = (n / p_sz, n % p_sz);
    (q * cumulated_pisano[p_sz as usize - 1] + cumulated_pisano[r as usize]) % MODULO
}
